### ISSUE TITLE
Adding load and save methods to WEASELClassifier

### ISSUE THREAD

**Issue by mohataher** [patrickzib/SFA#8]
After training, wouldn't it be a good idea to save a model so we could load it later?

Idealy, we need to add a static `load()` and `save()` methods to [`Classifier`](https://github.com/patrickzib/SFA/blob/master/src/main/java/sfa/classification/Classifier.java) so it would work with all classifiers implemented in the project.

I will be happy to help.

**Comment by patrickzib** [comment]
Sounds like a very good idea! 

I did implement something similar for the  [`SFATrie`].

How would you implement it? Using Java's serialization?

**Comment by mohataher** [comment]
Serialization is probably the best way to do it. 

I had a look at the code, specifically `WEASELModel`. Unfortunately, it's a challenge for us since `hppc` classes are not serializable. 

Now, we either find another alternative to serialization, modify hppc code, or change hppc dependency. I'll get back yo you with more drtails.

**Comment by patrickzib** [comment]
True, I just had a quick look and HPPC does not plan to support serialization anytime in future. There is a fork [HPPC-RT](https://github.com/vsonnier/hppcrt). But even they do not support it.

There seems to be support for [Jackson and HPPC](http://snipplr.com/view/63540/) though... not sure, if that is a feasible way to go.

I was wondering if there are other libraries that support serialization without the use of the Serializable-interface. But I didn't find a fast answer, yet...

**Comment by mohataher** [comment]
Thanks for looking into this. A good alternative is [Kryo](https://github.com/EsotericSoftware/kryo). It's a seriazable library that's much faster than Java Serialization and frequently used in big project inlcuding Apache Storm. I could have a quick test to see if it works with our collections and libraries. What do you think?

**Comment by patrickzib** [comment]
Sure! Kyro or [Fast-Serialization](https://github.com/RuedigerMoeller/fast-serialization) both sound good. Does Kyro work without the Serializable-interface?

**Comment by mohataher** [comment]
Yes, it doesn't require interface. I will give it a test on my fork and let you know

**Comment by patrickzib** [comment]
thanks a lot!

**Comment by mohataher** [comment]
It seems like all models are in `transformation` package. Right? I don't see an abstract `Model` that we could potentially add `load()` and `save()` methods. Shall I create one?

**Comment by patrickzib** [comment]
This is very good question! I am wondering what is part of the model and what is part of a classifier and what to store. I would say: the classifiers are the parameter-spaces to train and the model is the best parameter setup. Thus, my first guess is that the models are the right way to go. 

Yes, WEASELModel, BOSSModel and BOSSVSModel are all models. 
- BOSSModel and BOSSVSModel should be self-containing. 

- WEASELModel is not self-containing, as there is also the model from liblinear. So, I am not sure how to handle it best.

- To make things worse, the Shotgun*-Classifiers don't have a model yet. Instead the model is somewhat in [Classifier.Score](https://github.com/patrickzib/SFA/blob/master/src/main/java/sfa/classification/Classifier.java). 


Soo, one needs to:
- Create an abstract model-class with load and save functionality

- WEASELModel and BOSSModel should extend this abstract class (BOSSVS extends BOSS)

- I am still not sure how to handle liblinear though, which is needed for the WEASELClassifier

- Create ShotgunModel and ShotgunEnsembleModel-Classes that stores the best parameters for each.

**Comment by patrickzib** [comment]
I have started to separate concepts. But I will need some more refactoring. There are new classes now:

- ShotgunModel

- EnsembleModel

**Comment by mohataher** [comment]
Patrick, before we dive deep into that, we have to locate where `predict()` method is located. Saving and loading the model will serve one job; users need to use that method. If we know where it should be, we will know whether Classifier or Model are the right ones to use. 

Now, where should the method `predict` be located?

**Comment by patrickzib** [comment]
You are right. Definitely in the `Classifier`.

**Comment by mohataher** [comment]
Thanks for clarifying. Good effort in the past refactoring. Just out of curisousity, what's the difference between `WeaselScore` and `WASELModel`?

**Comment by patrickzib** [comment]
Hey @mohataher,

you are definitely asking the right questions! :) Both concepts `Model` and `Score` were mixed and in retroperspective I am very unhappy with the naming.

So, I have split both concepts in the master-branch and also altered the names of former `BOSSModel`, i.e., in package `transformation`. 

So, finally a `Model` contains the necessary classes and parameters to do the transformation. Every `Classifier` should now contain its `Model`, obtained by training. This model can be used for prediction and be serialized to disk.

**Comment by mohataher** [comment]
Thank you, Patrick. The new structure looks a lot better; moving all necessary variables (e.g., linear classifier in WEASEL) was a good move.

I had a deeper look at the code and now I see the method `predict()` containing evaluation of the results based on passed labels. I'd argue that's not what constitute a valid `predict()` method. Usually, they require data with no labels and return a list/array of predicted labels (Example [1](http://weka.sourceforge.net/doc.stable/weka/classifiers/trees/J48.html#classifyInstance-weka.core.Instance-), [2](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression.predict)) . 

To reflect on current code, [WEASELClassifier.predict()](https://github.com/patrickzib/SFA/blob/master/src/main/java/sfa/classification/WEASELClassifier.java#L177) should be renamed to `evalPredict()` and split to two methods:
1. `predict()`:public that takes `TimeSeries[]` and returns predicted labels. This will allow others to load the model and pass unknown time series for classification. So a null label should be handled with no errors.
1. `testEval()`:protected that takes `TimeSeries[]` (with labels in)and `predictedLabels[]` and returns `Predictions`. This one should probably be a method under `Classifier` class.

What do you think?

**Comment by patrickzib** [comment]
I do like the idea! I would rather use `score()` as the returned object is a `Score`. This name is also in line with sklearn. So a classifier would provide the methods:

```
  // @return The accuracy on the train-samples
  public Score fit(final TimeSeries[] trainSamples);

  // @return The predictions for each test-sample 
  public Predictions predict(final TimeSeries[] testSamples);

  // @return The accuracy for the test-samples 
  public Score score(final TimeSeries[] testSamples);

  // @return The accuracy on the test- and train-samples
  public Score eval(
      final TimeSeries[] trainSamples, final TimeSeries[] testSamples);
```

Is that what you have in mind?

**Comment by mohataher** [comment]
That looks quite good. Just noticed you pushed more code. I've just finished some code on my side to reflect some of these changes. Shall I send you a pull request?

**Comment by mohataher** [comment]
Just to move forward before we get into code, how about this small modification into your proposed signatures?

Those who call `predict()` won't need neither `Score` nor `Predictions`, they just need labels. 

```
// @return The accuracy on the train-samples
  public Score fit(final TimeSeries[] trainSamples);

  // @return The labels for each sample 
  public String[] predict(final TimeSeries[] samples);

  // @return Predictions; includes labels and overall accuracy for the test-samples 
  public Predictions score(final TimeSeries[] testSamples);

  // @return The accuracy on the test- and train-samples
  public Score eval(
      final TimeSeries[] trainSamples, final TimeSeries[] testSamples);
```

**Comment by mohataher** [comment]
Have a look at this commit [here](https://github.com/mohataher/SFA/blob/master/src/main/java/sfa/classification/WEASELClassifier.java#L116-L152) with the proposed change.

**Comment by patrickzib** [comment]
Looks great! Two thoughts of mine:

1) What about changing:

`private int evalTest(TimeSeries[] testSamples, String[] labels)`

to

`private Predictions evalLabels(TimeSeries[] samples, String[] labels)`

Then you could just do:
`return evalLabels(testSamples, labels);`

2) And I am thinking about moving from `String[] ` to ` Double[]` for the labels. But maybe that is a different aspect to consider in future.

**Comment by mohataher** [comment]
1. That's fine by me
1. Even better! That should give enough flexibility to users

**Comment by patrickzib** [comment]
regarding 1) do you want to prepare a pull request? or should I make the changes?

**Comment by mohataher** [comment]
Pull request is sent

**Comment by patrickzib** [comment]
Kryo seems to work greatly! Well done!

**Comment by mohataher** [comment]
Thank you. Glad this feature is added.
